swift/example_code/s3/basics/Tests/basics-tests/basics-tests.swift

```diff
@@ -0,0 +1,649 @@
+/*
+   Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+   SPDX-License-Identifier: Apache-2.0
+*/
+import XCTest
+import class Foundation.Bundle
+import Dispatch
+import ClientRuntime
+import Fakery
+@testable import ServiceHandler
+
+let DEBUG_SEM = false
+struct TestWaiter {
+    let sem: DispatchSemaphore
+    var name: String = "Unnamed"
+    var timeout: Double = 2.0
+
+    init() {
+        sem = DispatchSemaphore(value: 0)
+        if DEBUG_SEM == true {
+            print("Sem init: \(name)")
+        }
+    }
+
+    init(name: String, timeout: Double = 2.0) {
+        self.name = name
+        self.timeout = timeout
+
+        sem = DispatchSemaphore(value: 0)
+        if DEBUG_SEM == true {
+            print("Sem init: \(name)")
+        }
+    }
+
+    /// Signal the DispatchSemaphore, ending the lock.
+    public func signal() {
+        if DEBUG_SEM == true {
+            print("Sem signal: \(name)")
+        }
+        sem.signal()
+    }
+
+    /// Wait until the semaphore is unlocked, or until
+    /// the timeout period has elapsed, whichever comes
+    /// first.
```

Comment:
Wait until the semaphore is unlocked or the timeout period elapses, whichever comes first.